### PR TITLE
fix: #106 Supporting Environment Agnostic Stacks

### LINKED ISSUE #106: Supporting Environment Agnostic Stacks
This construct does not currently support AWS CDK's [environment agnostic stacks](https://docs.aws.amazon.com/cdk/v2/guide/environments.html). Because AWS offers Amazon Managed Blockchain in specific regions and availability zones, validation at the construct level was prioritized in conception, which causes AWS CDK apps that do not set a specific `region` in the environment variables section to fail at `region` validation.

### PR DESCRIPTION
**Pull request by trinitykey**
Fixes #106 

In order to support [Environment Agnostic Stacks](https://docs.aws.amazon.com/cdk/v2/guide/environments.html), we need to check whether or not the `region` is set as an environment variable in the CDK app. If no environment variables have been set for `region`, CDK assigns a [token](https://docs.aws.amazon.com/cdk/v2/guide/tokens.html) as a temporary placeholder until `cdk deploy` time. 

Based on the value of `region` in the CDK app's environment variables, whether it is a `string` representing the region or a CDK token representing a value to be received during `cdk deploy`, we can leave validation for either `cdk deploy` time or `cdk synth` time, respectively. 

If no `region` has been set in the environment variables, we assign a token to be resolved to both the `region` and default `availabilityZone` in the constructor. Validation of both will be performed at `cdk deploy`. 

If `region` has been set in the environment variables, we validate `region` and `availabilityZone` at `cdk synth` time using the validation methods in the `utilities` file. 

This solution provides support for native CDK environment agnostic stacks and validation at the construct level for better usability.

### CODE REVIEW

File: src/node.ts [hunk 1]
```diff
@@ -85,7 +85,12 @@ export class EthereumNode extends constructs.Construct {
 
     super(scope, id);
 
+
     const region = cdk.Stack.of(this).region;
+    const availabilityZone = cdk.Stack.of(this).availabilityZones;
+    // In order to support Environment Agnostic Stacks (https://docs.aws.amazon.com/cdk/v2/guide/environments.html),
+    // we need to check whether or not the region is set as an environment variable in the CDK app.
+    var tokenRegion = cdk.Token.isUnresolved(region);
```
Comment (lordjabez) [inline src/node.ts:93]: Let's be a bit more descriptive here with something like:

```
const regionInEnvironment = !cdk.Token.isUnresolved(region);
```

File: src/node.ts [hunk 2]
```diff
@@ -97,10 +102,21 @@ export class EthereumNode extends constructs.Construct {
     // will be populated with defaults when passed to the node constructor
     this.network = props.network ?? Network.MAINNET;
     this.instanceType = props.instanceType ?? InstanceType.BURSTABLE3_LARGE;
-    this.availabilityZone = props.availabilityZone ?? `${region}a`;
 
-    utilities.validateRegion(region);
-    utilities.validateAvailabilityZone(region, this.availabilityZone);
+
+    if (tokenRegion) {
```
Comment (lordjabez) [inline src/node.ts:107]: Here's how I'd suggest simplifying. See how the code is almost completely self-documenting now, no need for detailed comments.

```
// If no availability zone is provided, just use the first one.
this.region = props.region ?? region;
this.availabilityZone = props.availabilityZone ?? availabilityZone[0];

if (regionInEnvironment) {
    utilities.validateRegion(region);
    utilities.validateAvailabilityZone(region, this.availabilityZone);
}
```

File: src/node.ts [hunk 3]
```diff
@@ -85,7 +85,12 @@ export class EthereumNode extends constructs.Construct {
 
     super(scope, id);
 
+
     const region = cdk.Stack.of(this).region;
+    const availabilityZone = cdk.Stack.of(this).availabilityZones;
+    // In order to support Environment Agnostic Stacks (https://docs.aws.amazon.com/cdk/v2/guide/environments.html),
```
Comment (lordjabez) [inline src/node.ts:91]: I don't think the explanation here is needed when the variable is more clearly named.

File: src/node.ts [hunk 4]
```diff
@@ -97,10 +102,21 @@ export class EthereumNode extends constructs.Construct {
     // will be populated with defaults when passed to the node constructor
     this.network = props.network ?? Network.MAINNET;
     this.instanceType = props.instanceType ?? InstanceType.BURSTABLE3_LARGE;
-    this.availabilityZone = props.availabilityZone ?? `${region}a`;
 
-    utilities.validateRegion(region);
-    utilities.validateAvailabilityZone(region, this.availabilityZone);
+
+    if (tokenRegion) {
+      // If the region is not set as an environment variable and the availability zone is not included in node instantiation,
```
Comment (lordjabez) [inline src/node.ts:108]: Add this to the list of properties, might as well have region as well as AZ.

  /**
     * The Region in which the node will be created
     * @default - us-east-1
     */
  readonly region?: string;

**Review by lordjabez — CHANGES_REQUESTED** [review]
This is a clever solution, I like it!

Can you also consider a similar PR against the hyperledger contruct?

File: src/node.ts (hunk 3 above)
Comment (trinitykey) [reply src/node.ts]: Removed

File: src/node.ts (hunk 1 above)
Comment (trinitykey) [reply src/node.ts]: Changed!

File: src/node.ts (hunk 2 above)
Comment (trinitykey) [reply src/node.ts]: I made a small change here to accommodate an edge case: 

```
// If no availability zone is provided, use the first in the region.
if (regionInEnvironment) {
    utilities.validateRegion(region);
    this.availabilityZone = props.availabilityZone ?? `${region}a`;
    utilities.validateAvailabilityZone(region, this.availabilityZone);
} else {
    this.availabilityZone = props.availabilityZone ?? availabilityZones[0];
}
```

If `region` is specified in the environment variables, we want to assign the real availability zone value and then validate it, and if no availability zone is specified, assign the first in the region. If left unspecified, you wouldn't be able to assign the first token in the availability zone array AND validate at the `cdk synth` simultaneously.

File: src/node.ts (hunk 4 above)
Comment (trinitykey) [reply src/node.ts]: I tried adding `region` as a node property, but seeing it does not populate `managedblockchain.cfnNode` at construction and the construct does not validate beyond the value assigned `region = cdk.Stack.of(this).region`, I do not think adding it to the props provides any value for the end user. Let me know what you think!

Comment (lordjabez) [reply src/node.ts]: I wasn't suggesting it as an _input_, I was suggesting it as an _output_, only to go along with availability zone.

File: src/node.ts [hunk 5]
```diff
@@ -85,7 +85,10 @@ export class EthereumNode extends constructs.Construct {
 
     super(scope, id);
 
+
```
Comment (lordjabez) [inline src/node.ts:88]: Extra line break?

**Review by lordjabez — CHANGES_REQUESTED** [review]

**Pull request closed without merging**